src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts

```diff
@@ -390,3 +390,48 @@ export enum MarkdownLineType {
 	EVERY_LINE,
 	WRAPPED_ABOVE_AND_BELOW
 }
+
+export class TogglePreviewAction extends ToggleableAction {
+
+	private static readonly previewShowLabel = localize('previewShowLabel', "Show Preview");
+	private static readonly previewHideLabel = localize('previewHideLabel', "Hide Preview");
+	private static readonly baseClass = 'codicon';
+	private static readonly previewShowCssClass = 'split-toggle-on';
+	private static readonly previewHideCssClass = 'split-toggle-off';
+	private static readonly maskedIconClass = 'masked-icon';
+
+	constructor(
+		id: string, toggleTooltip: boolean, showPreview: boolean
+	) {
+		super(id, {
+			baseClass: TogglePreviewAction.baseClass,
+			toggleOffLabel: TogglePreviewAction.previewShowLabel,
+			toggleOffClass: TogglePreviewAction.previewShowCssClass,
+			toggleOnLabel: TogglePreviewAction.previewHideLabel,
+			toggleOnClass: TogglePreviewAction.previewHideCssClass,
+			maskedIconClass: TogglePreviewAction.maskedIconClass,
+			shouldToggleTooltip: toggleTooltip,
+			isOn: showPreview
+		});
+	}
+
+	public get previewMode(): boolean {
+		return this.state.isOn;
+	}
+	public set previewMode(value: boolean) {
+		this.toggle(value);
+	}
+
+	public run(context: any): Promise<boolean> {
+		let self = this;
+		return new Promise<boolean>((resolve, reject) => {
```

Comment:
Just make this async and get rid of the promise wrapping.

```
this.previewMode = !this.previewMode;
context.cellModel.showPreview = this.previewMode;
return true;
```

Reply:
Rewritten!